tutorials/haskell/ghc-optimization-and-fusion/tutorial.md

```diff
@@ -0,0 +1,1932 @@
+  never fire because `map2` may be inlined first. Of course `map2` won't be
+  ever inlined because it's self-recursive, but GHC can't figure that out
+  (yet).
+
+* `mapFB` is a helper function that takes consing function `c`, function we
+  want to apply to list `f` and lambda's head inside also binds `x` which is
+  an input value (`f` is applied to it) and `ys` which is the rest of the
+  list. The function's LHS has only two arguments to facilitate inlining in
+  our particular case (see fusion rules). Of course we want it to be
+  inlined, but only at the end because some rules match on it and they would
+  be broken if it were inlined too early.
+
+* Inlining is essential to all this stuff because it brings together
+  otherwise separate pieces of code and lets GHC manipulate them as a whole.
+
+* The `"map2"` and `"build/foldr2"` rewrite rules are familiar to us
+  already. `"mapFB"` is rather trivial. As I said previously, we have here
+  what is called “pair rules”, that is, the `"map2List"` rule rewrites
+  things back to simple `map2` if by the phase 1 fusion did not happen. This
+  is also why we have normal definition for `map2`, not `build (…)` one — if
+  fusion doesn't happen, `build`/`foldr` stuff actually makes things worse,
+  that's why it should be “a thing to try” for the compiler, not default
+  implementation.
+
+OK, let's see if it actually improves anything:
+
+```
+benchmarking fused2
+time                 107.5 ms   (103.8 ms .. 110.2 ms)
+                     0.998 R²   (0.995 R² .. 1.000 R²)
+mean                 107.3 ms   (104.6 ms .. 109.8 ms)
+std dev              3.768 ms   (2.519 ms .. 6.098 ms)
+
+Case                  Bytes  GCs  Check
+fused2          161,259,568  310  OK
+```
+
+Well, it's certainly better than the version without any fusion whatsoever,
+but still kinda sucks. What's the problem though?
+
+As I said inlining is essential in this sort of fusion business. Notice that
+`foldr2` won't be inlined and it won't be rewritten either (unlike `map2`).
+This is because `foldr` is self-recursive. Let's make it non-recursive and
+inline:
+
+```haskell
+foldr2 :: (a -> b -> b) -> b -> [a] -> b
+foldr2 f z = go
+  where
+    go []     = z
+    go (y:ys) = y `f` go ys
+{-# INLINE [0] foldr2 #-}
+```
+
+We specify phase 0 because we want GHC to inline it, but only after fusion
+has happened (remember that if we inlined it too early it would break our
+fusion rules and they wouldn't fire).
+
+Let's give it another shot:
+
+```
+benchmarking fused2
+time                 17.87 ms   (17.48 ms .. 18.33 ms)
+                     0.996 R²   (0.992 R² .. 0.998 R²)
+mean                 17.94 ms   (17.61 ms .. 18.42 ms)
+std dev              962.6 μs   (689.0 μs .. 1.401 ms)
+
+Case                  Bytes  GCs  Check
+fused2           96,646,160  153  OK
+```
+
+Nothing to be ashamed of, in fact, this is the same result we would get if
+we used `map` and `foldr` from `base`
```

Comment:
- In particular, `foldl` and `zip` cannot be fused efficiently when ~~are~~ written via `build` and `foldr`...
- all the details here~~,~~**.** As already mentioned...